knative-operator/pkg/common/telemetry/telemetry.go

```diff
@@ -0,0 +1,121 @@
+
+type Telemetry struct {
+	name string
+	stop chan struct{}
+	tc   *controller.Controller
+	// Protects from processing order, if true we should install telemetry
+	// if it is false we need to uninstall in the next delete stage.
+	// We start by assuming no telemetry is available.
+	shouldInstallTelemetry bool
+}
+
+func NewTelemetry(name string, mgr manager.Manager, objects []runtime.Object, api client.Client) (*Telemetry, error) {
+	t := &Telemetry{}
+	t.name = name
+	t.stop = make(chan struct{})
+	t.shouldInstallTelemetry = true
+	tc, err := newTelemetryController(name, objects, mgr, t, api)
+	if err != nil {
+		return nil, err
+	}
+	t.tc = tc
+	return t, nil
+}
+
+// TryStartTelemetry setups telemetry per component either Eventing, KnativeKafka or Serving.
+// When called it assumes that the component has status ready.
+func (t *Telemetry) TryStartTelemetry(api client.Client, mgr manager.Manager) error {
+	if t == nil {
+		return nil
+	}
+	if t.shouldInstallTelemetry {
+		log.Info("starting telemetry for:", "component", t.name)
+		errCh := make(chan error, 1)
+		// Initialize metrics before we start the corresponding controller.
+		// There is a tiny window to miss events here, but should be ok for telemetry purposes.
+		t.initializeAndUpdateMetrics(api)
+		// Start our controller in a goroutine so that we do not block.
+		go func() {
+			defer close(errCh)
+			// Block until our controller manager is elected leader. We presume our
+			// entire process will terminate if we lose leadership, so we don'telemetry need
+			// to handle that.
+			<-mgr.Elected()
+			// Start our controller. This will block until it is stopped
+			// or the controller returns a starting error.
+			if err := (*t.tc).Start(t.stop); err != nil {
+				log.Error(err, "cannot start telemetry controller for", "component", t.name)
+				errCh <- err
+			}
+		}()
+		select {
+		case err := <-errCh:
+			t.shouldInstallTelemetry = true
+			return err
+		case <-time.After(1 * time.Second): // assume no error, as controller has blocked
+		}
+		t.shouldInstallTelemetry = false
+	}
+	return nil
+}
+
+// TryStopTelemetry stops telemetry per component either Eventing, KnativeKafka or Serving
+// When called it assumes that we are reconciling a deletion event.
+func (t *Telemetry) TryStopTelemetry() {
+	if t == nil {
+		return
+	}
+	if !t.shouldInstallTelemetry {
+		log.Info("stopping telemetry for:", "component", t.name)
+		// Stop the telemetry controller
+		close(t.stop)
+		// Can'telemetry use a closed channel
+		t.stop = make(chan struct{})
+		t.shouldInstallTelemetry = true
+	}
+}
+
+// initializeAndUpdateMetrics is used for taking a global snapshot of metrics
+// before we start a telemetry controller. Cost should be low since we are fetching from cache.
+func (t *Telemetry) initializeAndUpdateMetrics(api client.Client) {
+	if t == nil {
+		return
+	}
+	switch t.name {
+	case "eventing":
+		pingSourceG = serverlessTelemetryG.WithLabelValues("source_ping")
```

Comment:
What are these label values for?

Reply:
It is a gauge vec where you have one vec and set a label to create a number of different time series based on the labels. It is standard prometheus approach instead of creating N different vecs.

Reply:
Ah I see, and we can't initialize them in the metrics code becauuuuse... that would make them all show 0 and we don't want that?

Reply:
Yes at least at the beginning we dont want if some component is never used or the operator is restarted, but it is of no harm to initialize them (I had it that way before). 
However, when we stop the controller we could remove the labels by calling `serverlessTelemetryG.DeleteLabelValues("...")`, I am still thinking about it, as discussed [here](https://godoc.org/github.com/prometheus/client_golang/prometheus#MetricVec.GetMetricWithLabelValues).:
```
Keeping the Metric for later use is possible (and should be considered if performance is critical), but keep in mind that Reset, DeleteLabelValues and Delete can be used to delete the Metric from the MetricVec. In that case, the Metric will still exist, but it will not be exported anymore, even if a Metric with the same label values is created later.
```
Right now between deletes/installs of Serving, Eventing while the serverless operator is running we keep the metrics.

Reply:
Honestly, I would not want to do that. What's the harm in reporting 0 sources in a cluster where there are... well... 0 sources? Seems like unnecessary complexity to me.